services/distribution/src/main/java/app/coronawarn/server/services/distribution/statistics/StatisticsToProtobufMapping.java

```diff
@@ -0,0 +1,133 @@
+   */
+  public StatisticsToProtobufMapping(DistributionServiceConfig distributionServiceConfig,
+      KeyFigureCardFactory keyFigureCardFactory,
+      JsonFileLoader jsonFileLoader) {
+    this.distributionServiceConfig = distributionServiceConfig;
+    this.keyFigureCardFactory = keyFigureCardFactory;
+    this.jsonFileLoader = jsonFileLoader;
+  }
+
+
+  /**
+   * Create protobuf statistic object from raw JSON statistics.
+   *
+   * @return Statistics protobuf statistics object.
+   */
+  @Bean
+  public Statistics constructProtobufStatistics() {
+    String content = this.jsonFileLoader.getContent();
+    List<StatisticsJsonStringObject> jsonStringObjects = SerializationUtils
+        .deserializeJson(content, typeFactory -> typeFactory
+            .constructCollectionType(List.class, StatisticsJsonStringObject.class));
+
+    StatisticsJsonValidator validator = new StatisticsJsonValidator();
+    jsonStringObjects = new ArrayList<>(validator.validate(jsonStringObjects));
+
+    return Statistics.newBuilder()
+        .addAllCardIdSequence(getAllCardIdSequence())
+        .addAllKeyFigureCards(buildAllKeyFigureCards(jsonStringObjects))
+        .build();
+  }
+
+  private List<Integer> getAllCardIdSequence() {
+    String[] idSequenceArray = distributionServiceConfig.getCardIdSequence()
+        .replace("[", "").replace("]", "")
+        .split(",");
+    List<Integer> idIntegerSequence = new ArrayList<>();
+    for (String id : idSequenceArray) {
+      idIntegerSequence.add(Integer.parseInt(id));
+    }
+    return idIntegerSequence;
+  }
+
+  private LocalDate effectiveDateStringToLocalDate(String effectiveDate) {
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+    return LocalDate.parse(effectiveDate, formatter);
+  }
+
+  @SuppressWarnings("checkstyle:Indentation")
+  private List<KeyFigureCard> buildAllKeyFigureCards(List<StatisticsJsonStringObject> jsonStringObjects) {
+    Map<Integer, KeyFigureCard> figureCardMap = new HashMap<>();
+
+    figureCardMap.put(INFECTIONS_CARD_ID, null);
+    figureCardMap.put(INCIDENCE_CARD_ID, null);
+    figureCardMap.put(KEY_SUBMISSION_CARD_ID, null);
+    figureCardMap.put(FOURTH_CARD_ID, null);
+
+    List<StatisticsJsonStringObject> orderedList = jsonStringObjects.stream()
+        .sorted(Comparator.comparing(a -> effectiveDateStringToLocalDate(a.getEffectiveDate())))
+        .collect(Collectors.toList());
+    Collections.reverse(orderedList);
+
+    for (var stat : orderedList) {
+      getAllCardIdSequence().forEach(id -> {
+        if (figureCardMap.get(id) == null) {
+          KeyFigureCard card;
+          try {
+            card = keyFigureCardFactory.createKeyFigureCard(stat, id);
+            figureCardMap.put(id, card);
+          } catch (MissingPropertyException ex) {
+            logger.warn(ex.getMessage());
+          }
+        }
+      });
+
+      if (figureCardMap.values().stream().allMatch(Objects::nonNull)) {
```

Comment:
What is the purpose of checking whether values are null after the perivous loop? If no card was created with the previous stat than it should stop? Should we throw an exception ?

Reply:
The list of statistics is ordered by date, and if those 4 cards(INFECTIONS_CARD, INCIDENCE_CARD, KEY_SUBMISSION,FOURTH_CARD) are already populated with values from the most recent statistics entries from the json file, then we should not look into the older ones, so we stop the loop.

Reply:
It's checking for `nonNull`. The loop is going through the most recent record until the oldest, as Ioan mentioned. If half-way through we already have all four cards created, we can skip break out of the loop and return all `KeyFigureCard` objects.

Reply:
Sorry, my bad. I misread isNull instead of nonNull